[DCOS-59566] Migrate tests from Helm to KUDO Spark

rpalaznik opened the pull request:
### What changes were proposed in this pull request?
Resolves [DCOS-59566](https://jira.mesosphere.com/browse/DCOS-59566)

The operator fixture installs the operator with KUDO client instead of using external helm charts. Before the installation a clean-up procedure (identical to `scripts/remote_operator.py`) is performed in the namespace. 

### How were the changes tested?
Manually with `make test`

alembiewski commented:
Thanks @rpalaznik!
Tried to run it locally and it looks like `TestJobSubmission` has failed with because of:
```
+ kubectl kudo --namespace kudo-spark-operator-testing install /kudo-spark-operator/kudo-operator -p operatorImageName=mesosphere/kudo-spark-operator -p operatorVersion=4f7498c063d7d3ca6f13c49b727863100df18dbb
E1014 14:15:21.054475    3540 request.go:853] Unexpected error when reading response body: &http.httpError{err:"net/http: request canceled (Client.Timeout exceeded while reading body)", timeout:true}
Error: installing OperatorVersion CRD for operator: spark: installing spark-operatorversion.yaml: installing OperatorVersion: Unexpected error &http.httpError{err:"net/http: request canceled (Client.Timeout exceeded while reading body)", timeout:true} when reading response body. Please retry.
```
Full log is [attached](https://github.com/mesosphere/kudo-spark-operator/files/3724833/test_output.log)

rpalaznik commented:
> Tried to run it locally and it looks like `TestJobSubmission` has failed with because of:
> 
> ```
> + kubectl kudo --namespace kudo-spark-operator-testing install /kudo-spark-operator/kudo-operator -p operatorImageName=mesosphere/kudo-spark-operator -p operatorVersion=4f7498c063d7d3ca6f13c49b727863100df18dbb
> E1014 14:15:21.054475    3540 request.go:853] Unexpected error when reading response body: &http.httpError{err:"net/http: request canceled (Client.Timeout exceeded while reading body)", timeout:true}
> Error: installing OperatorVersion CRD for operator: spark: installing spark-operatorversion.yaml: installing OperatorVersion: Unexpected error &http.httpError{err:"net/http: request canceled (Client.Timeout exceeded while reading body)", timeout:true} when reading response body. Please retry.
> ```

Yep. `kudo` likes to fail with a timeout error if your internet connection is not perfect. I had to develop and test this PR in a remote AWS machine in us-west region because of this.

rpalaznik commented on images/builder/Dockerfile:
```diff
@@ -7,5 +7,6 @@ RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.16.0/b
     mv linux-amd64/helm /usr/local/bin/helm && \
-    rm helm-v2.14.3-linux-amd64.tar.gz*
-
+    rm helm-v2.14.3-linux-amd64.tar.gz* && \
```
Thanks. I removed it.

samvantran commented on tests/utils/spark_operator.go:
```diff
@@ -1,135 +1,158 @@
-	pods, err := spark.Clients.CoreV1().Pods(spark.Namespace).List(v1.ListOptions{LabelSelector: "app.kubernetes.io/name=sparkoperator"})
+func (spark *SparkOperatorInstallation) CleanUp() {
+	// So far multiple Spark operator instances in one namespace is not a supported configuration, so whole namespace will be cleaned
+	log.Infof("Uninstalling _all_ kudo spark operator instances and versions from %s", spark.Namespace)
```
```suggestion
	log.Infof("Uninstalling ALL kudo spark operator instances and versions from %s", spark.Namespace)
```
Or does `_x_` actually get printed in italics?

samvantran commented on tests/utils/spark_operator.go:
```diff
@@ -1,135 +1,158 @@
 
-	return pods.Items[0], nil
+	DeleteResource(spark.Namespace, "operator.kudo.dev", "spark")
+	//DropNamespace(spark.Clients, spark.Namespace)
```
still valid or should we delete?

rpalaznik replied:
`remove_operator.py` doesn't remove the namespace, but on second though I think it's valid to do so in the tests. I've un-commented the line.

samvantran commented on tests/utils/utils_test.go:
```diff
@@ -22,20 +22,14 @@ func TestClientGo(t *testing.T) {
+	})
 	defer os.Remove(tmpFilePath)
 
 	log.Infof("Created a temp file at %s", tmpFilePath)
```
should there be an assert in this test? Not sure what it's actually testing.

rpalaznik replied:
This whole file is not included in test runs currently, i.e. not executed with `make test`. I'm using it mostly for debugging various utility methods. Later, when there are more complex logic in the tests, I plan to either turn it into proper tests or remove completely.

samvantran commented:
FTR, tests passed locally against my own konvoy cluster. But I noticed a few things that might improve the stdout/err:

- Can we do a check to see if kudo is already installed or pass the `--client-only` flag ? I saw this error:
  - ```
    Using namespace 'kudo-spark-operator-testing' for installation
    + kubectl kudo init
    $KUDO_HOME has been configured at /root/.kudo
    Warning: KUDO is already installed in the cluster.
    (Use --client-only to suppress this message)
    Error: error installing: services "kudo-controller-manager-service" already exists
    ```
    and then it goes on to print the `kudo help` info. This happens multiple times per run.
- nit: when cleaning up after the test run I see we try to delete an empty operatorVersion 
  - ```
    INFO[2019-10-14T20:16:08Z] >/usr/local/bin/kubectl [kubectl delete operatorversion.kudo.dev  --namespace kudo-spark-operator-testing]
    error: resource name may not be empty
    INFO[2019-10-14T20:16:09Z] >/usr/local/bin/kubectl [kubectl delete operatorversion.kudo.dev spark-0.0.1 --namespace kudo-spark-operator-testing]
    operatorversion.kudo.dev "spark-0.0.1" deleted

rpalaznik commented:
@akirillov, @samvantran could you please review this PR again. I've implemented the rest of your suggestions:

- Stopped using `install_operator.sh` script.
- One suite-scoped attempt to install KUDO which won't throw an error if it's already installed
- Can pass params to `kubectl-kudo install ...` for future tests
- Hardcoded instance name
- Fixed empty operation version bug
- Updated tooling submodule and CI is :heavy_check_mark: now

akirillov commented on tests/utils/spark_operator.go:
```diff
@@ -1,135 +1,158 @@
-	if err != nil {
-		log.Fatal(err)
-		panic(err)
+func (spark *SparkOperatorInstallation) InstallSparkOperatorWithParams(params map[string]string) error {
```
does it make sense to move `params map[string]string` into `SparkOperatorInstallation` so we have all the configuration encapsulated in one place?

rpalaznik replied:
It does, thanks! I've applied your suggestion.

akirillov commented:
Looks like the PR description doesn't reflect the implementation anymore, shall we update it?

VladimirKravtsov commented on tests/utils/spark_operator.go:
```diff
@@ -1,135 +1,158 @@
 
-func (spark *SparkOperatorInstallation) CleanUp() {
-	uninstallSparkOperatorWithHelm(spark.Namespace)
+	installScript := exec.Command("bash", path.Join(TestDir, "../scripts/install_operator.sh"))
```
Hi @akirillov 
> So it would be great to have the following:
> 
> * migrate off from using `install_operator.sh` script and
>   

I also agree that once `kudo install <operater>` will be good enough to install fully working operator(with ability to enable/disable modes by a parameters specifying) it should be used in tests to spin up the operator. However, as for now, we have a couple of workarounds accumulated in `install_operator.sh`, `remove_operator.sh` like monitoring service, crds installation etc. And I think it was a bit early to migrate from it now. Because currently we are often make changes to the `install_operator.sh` and it will not be reflected to here. I mean we should maintain these workarounds in two places now. Just IMO.
